lib/data_freshness/main.py

```diff
@@ -0,0 +1,160 @@
+                dict(transaction, institution_id=institution_id)
+            )
+
+    # STEP a4: Here we sort by timestamp so latest transactions are on top and keep the first entry
+    transactions_flattened = sorted(
+        transactions_flattened, key=lambda x: x["ts"], reverse=True
+    )
+
+    # STEP a5: Get timestamp of first transaction
+    if transactions_flattened is None:
+        transactions_max_ts = 0
+    else:
+        transactions_max_ts = transactions_flattened[0]["ts"]
+
+    # STEP b2: get a list of all balances for each institution
+    balance_inst_coroutines = []
+    for inst in institutions:
+        balance_inst_coroutines.append(
+            api_client.balances.get(
+                user_uuid=user_uuid,
+                institution_id=inst.institution_id,
+                utc_starttime=utc_starttime,
+                utc_endtime=utc_endtime,
+                page=1,
+            )
+        )
+
+    balances_per_institution = await asyncio.gather(*balance_inst_coroutines)
+
+    # STEP b3: flatten all balances from all institutions
+    balances_flattened = []
+    for ix, balances in enumerate(balances_per_institution):
+        institution_id = institutions[ix].institution_id
+        for balance in balances:
+            balances_flattened.append(dict(balance, institution_id=institution_id))
+
+    # STEP b4: Here we sort by timestamp so latest transactions are on top and keep the first entry
+    balances_flattened = sorted(balances_flattened, key=lambda x: x["ts"], reverse=True)
+
+    # STEP b5: Get timestamp of first transaction
+    if balances_flattened is None:
+        balances_max_ts = 0
+    else:
+        balances_max_ts = balances_flattened[0]["ts"]
+
+    # STEP c2: get a list of all alerts for each institution
+    alerts_inst_coroutines = []
+    for inst in institutions:
+        alerts_inst_coroutines.append(
+            api_client.balances.get(
+                user_uuid=user_uuid,
+                institution_id=inst.institution_id,
+                utc_starttime=utc_starttime,
+                utc_endtime=utc_endtime,
+                page=1,
+            )
+        )
+
+    alerts_per_institution = await asyncio.gather(*alerts_inst_coroutines)
+
+    # STEP c3: flatten all balances from all institutions
+    alerts_flattened = []
+    for ix, alerts in enumerate(alerts_per_institution):
+        institution_id = institutions[ix].institution_id
+        for alert in alerts:
+            alerts_flattened.append(dict(alert, institution_id=institution_id))
+
+    # STEP c4: Here we sort by timestamp so latest transactions are on top and keep the first entry
+    alerts_flattened = sorted(alerts_flattened, key=lambda x: x["ts"], reverse=True)
+
+    # STEP c5: Get timestamp of first transaction
+    if alerts_flattened is None:
+        alerts_max_ts = 0
+    else:
+        alerts_max_ts = alerts_flattened[0]["ts"]
+
+    # STEP 6: Finally, we can get the minimum of the data freshness (in days)
```

Comment:
Re: step numbers above, I think it's a good sign to restructure if the step numbers are entering a second layer/level

One way of restructuring I can think of is having the steps mirror the existing features ->


Step 1: fetches list of institutions belonging to the user (perfect as is)
Step 2: Fetches transactions, balances and alerts for each institution -> You can have lines `68 through 80` and `99 through 111` clubbed near line 41
Step 3: Computing freshness -> All code related to computing freshness can live here, since all the `fetching` would have happened in `step 2` itself

Please let me know if ^^ is not clear 😅